apps/api/src/app/groups/groups.service.ts

```diff
@@ -283,23 +269,54 @@ export class GroupsService {
         return cachedGroup.generateMerkleProof(memberIndex)
     }
 
+    private async _cacheGroups() {
+        const groups = await this.getAllGroups()
+
+        /* istanbul ignore next */
+        for (const group of groups) {
+            const cachedGroup = new CachedGroup(group.id, group.treeDepth)
+
+            cachedGroup.addMembers(group.members.map((m) => m.id))
+
+            this.cachedGroups.set(group.id, cachedGroup)
+        }
+
+        Logger.log(`GroupsService: ${groups.length} groups have been cached`)
+    }
+
+    /**
+     * If the off-chain groups roots don't match the contract's ones, it updates them.
+     */
+    /* istanbul ignore next */
+    private async _syncContractGroups() {
+        const contractGroups = await this.bandadaContract.getGroups()
+        const fingerprints = new Set(
+            contractGroups.map(({ fingerprint }) => fingerprint.toString())
+        )
+
+        for (const [, group] of this.cachedGroups) {
+            if (!fingerprints.has(group.root.toString())) {
```

Comment:
This will add groups that are missing in the contract; how about also updating if the value is not same in the contract? Its less likely to happen, but in case where a regular group update failed to update the contract, we can fix it during next startup.

Reply:
Oh you are right, it works that way. I interpreted the code wrong earlier.